### PR TITLE
DSL2: Dev hostremoval

### PR DESCRIPTION
**Pull request by aidaanva**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/local/host_removal.nf [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    meta_out.sample_id          = meta_fastqs.sample_id
+    meta_out.library_id         = meta_fastqs.library_id
+    meta_out.lane               = meta_fastqs.lane
+    meta_out.colour_chemistry   = meta_fastqs.colour_chemistry
+    meta_out.single_end         = meta_fastqs.single_end
+    meta_out.strandedness       = meta_fastqs.strandedness
+    meta_out.damage_treatment   = meta_fastqs.damage_treatment
+    meta_out.reference          = meta.reference
+
+
```
Comment (jfy133) [inline modules/local/host_removal.nf:35]:
```suggestion
```

File: modules/local/host_removal.nf [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/mulled-v2-adbddd275b72bafc166f5c77e6b1a8c3d200cfe7:2c601eb41a331051f5d90b147919a69ac4e17e19-0' :
+        'quay.io/biocontainers/mulled-v2-adbddd275b72bafc166f5c77e6b1a8c3d200cfe7:2c601eb41a331051f5d90b147919a69ac4e17e19-0' }"
+
+    input:
+    // [ meta_bam, bam, meta_fastqs, fastqs ]
+    tuple val(meta), path(bam), path(bai), val(meta_fastqs), path(fastqs)
+
+    output:
+    tuple val(meta_out), path("*.fq.gz"), emit: fastqs
```
Comment (jfy133) [inline modules/local/host_removal.nf:16]:
```suggestion
    tuple val(meta), path("*.fq.gz"), emit: fastqs
```

**Review by jfy133 — COMMENTED** [review]
We don't use the resulting files in any downstream step, so we can arbitrarily just pick the BAM `meta`

File: modules/local/host_removal.nf [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    meta_out.lane               = meta_fastqs.lane
+    meta_out.colour_chemistry   = meta_fastqs.colour_chemistry
+    meta_out.single_end         = meta_fastqs.single_end
+    meta_out.strandedness       = meta_fastqs.strandedness
+    meta_out.damage_treatment   = meta_fastqs.damage_treatment
+    meta_out.reference          = meta.reference
+
+
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta_out.id}_${meta_out.reference}"
```
Comment (aidaanva) [inline modules/local/host_removal.nf:37]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}_${meta.reference}"
```

File: conf/modules.config [hunk 4]
```diff
@@ -382,6 +382,18 @@ process {
         ]
     }
 
+    withName: HOST_REMOVAL {
+        ext.args= {[ "-m ${params.hostremoval_mode}",
+        "${meta_fastqs.single_end}" == false && "!${params.preprocessing_skippairmerging}" ? "-merged" : ""
```
Comment (jfy133) [inline conf/modules.config:387]: Jsut to confirm: mMerged also applies for single end data?

File: docs/output.md [hunk 5]
```diff
@@ -145,6 +145,23 @@ The resulting FASTQ files will only be present in your results directory if you
...
+<summary>Output files</summary>
+
+- `host_removal/`
+
+  - `*.fq.gz`: Fastq files containing only reads that did not map
+    to the reference genome.
+
+</details>
+
+nf-core/eager host removal step removes any reads from the original input fastq files that mapped to the provided reference. This step is performed by an a costume python script [extract_map_reads.py](https://github.com/nf-core/eager/blob/master/bin/extract_map_reads.py).
```
Comment (jfy133) [inline docs/output.md:160]:
```suggestion
nf-core/eager host removal step removes any reads from the original input fastq files that mapped to the provided reference. This step is performed by a custom python script [extract_map_reads.py](https://github.com/nf-core/eager/blob/master/bin/extract_map_reads.py).
```

File: docs/output.md [hunk 6]
```diff
@@ -145,6 +145,23 @@ The resulting FASTQ files will only be present in your results directory if you
...
+
+The resulting fastq files can be used for publication purposes where
+the host DNA should not be made available for ethical reasons (for example when you do not have permission to analyse the host DNA but analysis on the metagenomic content were agreed) or for due to identifiability reasons (for example, for individualsthat died in recent times and could be identified based on their DNA alone).
```
Comment (jfy133) [inline docs/output.md:163]:
```suggestion
The resulting FASTQ files can be used for publication purposes where
the host DNA should not be made available for ethical reasons (for example when you do not have permission to analyse the host DNA but only analysis on the metagenomic content were agreed) or for due to identifiability reasons (for example, for individuals that died in recent times and could be identified based on their DNA alone).

Alternatively you could use it for manual metagenomic screen outside of nf-core/eager (e.g. when your preferred taxonomic classifier isn't supported by nf-core/eager).
```

File: modules/local/host_removal.nf [hunk 7]
```diff
@@ -0,0 +1,48 @@
+process HOST_REMOVAL {
+    tag "$meta.id"
+    label 'process_medium'
+
+    //TODO:check if correct conda and containers used
```
Comment (jfy133) [inline modules/local/host_removal.nf:5]: ??? Is this checked now?

File: modules/local/host_removal.nf [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    """
+    extract_map_reads.py \\
+        $args \\
+        $outputR2 \\
+        -of ${prefix}_hostremoved.r1.fq.gz \\
+        -t $task.cpus \\
+        $bam \\
+        $fastqs_input
+
```
Comment (jfy133) [inline modules/local/host_removal.nf:38]:
```suggestion
```

File: nextflow_schema.json [hunk 9]
```diff
@@ -741,6 +741,28 @@
...
+                    "description": "Turn on per-library creation pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)",
+                    "help_text": "Create pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)",
+                    "fa_icon": "fas fa-power-off"
+                },
+                "hostremoval_mode": {
+                    "type": "string",
+                    "default": "remove",
+                    "description": "Host removal mode. Remove mapped reads completely from FASTQ (remove) or just mask mapped reads sequence by N (replace).",
+                    "help_text": "Read removal mode. Remove mapped reads completely ('remove') or just replace mapped reads sequence by N ('replace')\n\n> Modifies extract_map_reads.py parameter: -m",
+                    "fa_icon": "fas fa-plane-slash"
```
Comment (jfy133) [inline nextflow_schema.json:762]: Missing the enum of options - in schema builder press 'cog' next to the entry and fill in the pipe separate d list

File: subworkflows/local/endorspy.nf [hunk 10]
```diff
@@ -0,0 +1,47 @@
+//
+// Calculate different percent on target as well as complexity from samtools flagstats files
+//
+
+include { ENDORSPY_COMPLETE              } from '../../modules/nf-core/endorspy'
+include { ENDORSPY_QUALITY_FILTER        } from '../../modules/nf-core/endorspy'
+include { ENDORSPY_RAW                   } from '../../modules/nf-core/endorspy'
+include { ENDORSPY_NOQF_DEDUP                   } from '../../modules/nf-core/endorspy'
```
Comment (jfy133) [inline subworkflows/local/endorspy.nf:8]: Is endorSpy meant to be in here? Not reviewing for now...

**Review by jfy133 — CHANGES_REQUESTED** [review]

File: conf/modules.config (hunk 4 above)
Comment (aidaanva) [reply conf/modules.config:387]: @jfy133: The merged flag only applies for pair end reads that have been merged. So in here I am checking if meta_fastqs is paired_end and you haven't set the skip pair merging, meaning they were merged

Comment (jfy133) [reply conf/modules.config:405]: :+1:

File: docs/output.md [hunk 11]
```diff
@@ -160,6 +160,25 @@ The resulting FASTQ files will only be present in your results directory if you
...
   [Bowtie 2](https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml) is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s of characters to relatively long (e.g. mammalian) genomes. Bowtie 2 indexes the genome with an FM Index (based on the Burrows-Wheeler Transform or BWT) to keep its memory footprint small and supports gapped, local, and paired-end alignment modes.
 
+### Host Removal
+
+<details markdown="1">
+<summary>Output files</summary>
+
+- `host_removal/`
+
+  - `*.fq.gz`: Fastq files containing only reads that did not map
```
Comment (jfy133) [inline docs/output.md:170]:
```suggestion
  - `*.fq.gz`: FASTQ files containing only reads that did not map
```

File: docs/output.md [hunk 12]
```diff
@@ -160,6 +160,25 @@ The resulting FASTQ files will only be present in your results directory if you
...
+    to the reference genome.
+
+</details>
+
+nf-core/eager host removal step removes any reads from the original input fastq files that mapped to the provided reference. This step is performed by a custom python script [extract_map_reads.py](https://github.com/nf-core/eager/blob/master/bin/extract_map_reads.py).
```
Comment (jfy133) [inline docs/output.md:175]:
```suggestion
nf-core/eager host removal step removes any reads from the original input FASTQ files that mapped to the provided reference. This step is performed by a custom python script [extract_map_reads.py](https://github.com/nf-core/eager/blob/master/bin/extract_map_reads.py).
```

File: docs/output.md [hunk 13]
```diff
@@ -160,6 +160,25 @@ The resulting FASTQ files will only be present in your results directory if you
...
+
+The resulting FASTQ files can be used for publication purposes where
+the host DNA should not be made available for ethical reasons (for example when you do not have permission to analyse the host DNA but only analysis on the metagenomic content were agreed) or for due to identifiability reasons (for example, for individuals that died in recent times and could be identified based on their DNA alone).
```
Comment (jfy133) [inline docs/output.md:178]:
```suggestion
The resulting FASTQ files can be used for publication purposes where the host DNA should not be made available for ethical reasons (for example when you do not have permission to analyse the host DNA but only analysis on the metagenomic content were agreed) or for due to identifiability reasons (for example, for individuals that died in recent times and could be identified based on their DNA alone).
```

File: modules/local/host_removal.nf [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    label 'process_medium'
+
+    conda "bioconda::xopen=1.1.0 bioconda::pysam=0.16.0"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/mulled-v2-adbddd275b72bafc166f5c77e6b1a8c3d200cfe7:2c601eb41a331051f5d90b147919a69ac4e17e19-0' :
+        'quay.io/biocontainers/mulled-v2-adbddd275b72bafc166f5c77e6b1a8c3d200cfe7:2c601eb41a331051f5d90b147919a69ac4e17e19-0' }"
+
+    input:
+    // [ meta_bam, bam, meta_fastqs, fastqs ]
+    tuple val(meta), path(bam), path(bai), val(meta_fastqs), path(fastqs)
```
Comment (jfy133) [inline modules/local/host_removal.nf:12]:
```suggestion
    tuple val(meta), path(bam), path(bai), val(meta_fastqs), path(fastqs) // [ meta_bam, bam, meta_fastqs, fastqs ]
```

File: modules/local/host_removal.nf [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}_${meta.reference}"
+    def VERSION_PYSAM = '0.16.0'
+    def VERSION_XOPEN = '1.1.0'
+    def fastqs_input = [fastqs].flatten().sort().size() > 1 ? "${fastqs[0]} -rev ${fastqs[1]}" : "${fastqs[0]}"
+    def outputR2 = [fastqs].flatten().sort().size() > 1 ? "-or ${prefix}_hostremoved.r2.fq.gz" : ''
```
Comment (jfy133) [inline modules/local/host_removal.nf:27]: All variables should be `snake_case`

**Review by jfy133 — COMMENTED** [review]
Couple of minor things, but looks overall pretty good.

Could you list what tests you've checked to account for all configuration options

File: nextflow_schema.json (hunk 9 above)
Comment (aidaanva) [reply nextflow_schema.json]: done

File: modules/local/host_removal.nf (hunk 7 above)
Comment (aidaanva) [reply modules/local/host_removal.nf]: yes

**Comment by aidaanva** [comment]
Added all the manual test done as agreed in nf-thusiasts

File: subworkflows/local/hostremoval.nf
Comment (TCLamnidis) [inline subworkflows/local/hostremoval.nf:1]: Does this actually work? Won't it mix up the SWF and MODULE if they are both named the same? ?_?

Comment (TCLamnidis) [reply subworkflows/local/hostremoval.nf:1]: SWF needs renaming anyway 😇

File: workflows/eager.nf [hunk 16]
```diff
@@ -232,6 +232,34 @@ workflow EAGER {
...
+                                                new_meta = meta.clone().findAll{ it.key !in [ 'single_end', 'reference' ] }
+                                                [ new_meta, meta, bam, bai ]
+                                                }
+        ch_fastqs_for_hostremoval= INPUT_CHECK.out.fastqs.map{
+                                                        meta, fastqs ->
+                                                        new_meta = meta.clone().findAll{ it.key !in [ 'lane', 'colour_chemistry', 'single_end' ] }
+                                                        [ new_meta, meta, fastqs ]
+                                                    }
+
+        ch_for_hostremoval = ch_bam_for_hostremoval.join(ch_fastqs_for_hostremoval)
```
Comment (TCLamnidis) [inline workflows/eager.nf:252]: I'd rename `ch_for_hostremoval` to `ch_input_for_hostremoval` personally, but 🤷

**Review by TCLamnidis — CHANGES_REQUESTED** [review]
- [x] Remove all `.dump()` calls
- [ ] `manual_tests.md`: Add Expectations and explain what you looked for to ensure it worked.
- [ ] Rename Subworkflow to be an action (`REMOVE_HOST`) 


Sometimes you use `host_removal` (e.g. for output dirs) and sometimes `hostremoval` (e.g. for parameter names. Could you homogenise?

File: nextflow_schema.json [hunk 17]
```diff
@@ -792,6 +792,29 @@
...
+                    "type": "boolean",
+                    "description": "Turn on per-library creation pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)",
+                    "help_text": "Create pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)",
+                    "fa_icon": "fas fa-power-off"
+                },
+                "hostremoval_mode": {
+                    "type": "string",
+                    "default": "remove",
+                    "description": "Host removal mode. Remove mapped reads completely from FASTQ (remove) or just mask mapped reads sequence by N (replace).",
+                    "help_text": "Read removal mode. Remove mapped reads completely ('remove') or just replace mapped reads sequence by N ('replace')\n\n> Modifies extract_map_reads.py parameter: -m",
```
[the remark this answers is not in the document]
Comment (aidaanva) [reply nextflow_schema.json:848]: Will remove everything before `Modifies parameter...`

File: docs/development/manual_tests.md
[the remark this answers is not in the document]
Comment (aidaanva) [reply docs/development/manual_tests.md:1]: done!

**Review by TCLamnidis — CHANGES_REQUESTED** [review]
Just two minor things! 😄  😇 
- [ ] Manual tests still lack `Expect:` detailing the expected output that as verified as present.
- [ ] Could you add some comments explaining the channel manipulations in `eager.nf`? Ideally, explain why each aspect of the map is dropped. The reason is clear currently, but will be useful for posterity.

Also @jfy133, Currently I do not see a change in the `meta.single_end` when merging PE data during preprocessing. Should that maybe be happening?

**Review by TCLamnidis — APPROVED** [review]
LGTM ❤️

File: docs/output.md [hunk 18]
```diff
@@ -160,6 +160,24 @@ The resulting FASTQ files will only be present in your results directory if you
...
+<summary>Output files</summary>
+
+- `host_removal/`
+
+  - `*.fq.gz`: FASTQ files containing only reads that did not map
+    to the reference genome.
+
+</details>
+
+nf-core/eager host removal step removes any reads from the original input FASTQ files that mapped to the provided reference. This step is performed by a custom python script [extract_map_reads.py](https://github.com/nf-core/eager/blob/master/bin/extract_map_reads.py).
```
Comment (jfy133) [inline docs/output.md:175]:
```suggestion
The nf-core/eager host removal step removes any reads from the original input FASTQ files that mapped to the provided reference. This step is performed by a custom python script [extract_map_reads.py](https://github.com/nf-core/eager/blob/master/bin/extract_map_reads.py).
```

File: nextflow_schema.json [hunk 19]
```diff
@@ -829,6 +829,29 @@
...
         },
+        "host_removal": {
+            "title": "Host Removal",
+            "type": "object",
+            "description": "",
+            "default": "",
+            "properties": {
+                "run_host_removal": {
+                    "type": "boolean",
+                    "description": "Turn on per-lane creation pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)",
```
Comment (jfy133) [inline nextflow_schema.json:840]:
```suggestion
                    "description": "Turn on per-lane creation  of pre-adapter removed and/or merged FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)",
```

File: nextflow_schema.json [hunk 20]
```diff
@@ -829,6 +829,29 @@
...
+                    "help_text": "Create pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy-sensitive non-host data)",
```
Comment (jfy133) [inline nextflow_schema.json:841]:
```suggestion
                    "help_text": "Recreates pre-adapter and/or merged FASTQ files but without reads that mapped to reference (e.g. for public upload of privacy-sensitive non-host data)",
```

File: nextflow_schema.json [hunk 21]
```diff
@@ -829,6 +829,29 @@
...
+                    "fa_icon": "fas fa-power-off"
+                },
+                "host_removal_mode": {
+                    "type": "string",
+                    "default": "remove",
+                    "description": "Host-mapped read removal mode. Remove mapped reads completely from FASTQ (remove) or just mask the sequence of mapped reads with N (replace).",
```
Comment (jfy133) [inline nextflow_schema.json:847]:
```suggestion
                    "description": "Host-mapped read removal mode. Remove mapped reads completely from FASTQ (remove) or just mask the host sequence of mapped reads with N (replace).",
```

Comment (jfy133) [reply nextflow_schema.json:847]: Could add a extended `help_text` saying, can be useful to set when you want to make sure you have the same number of reads as the input data to your host-removed publically uploaded data, but still removing any host information.

**Review by jfy133 — APPROVED** [review]
A few final docs changes but otherwise LGTM

**Pull request merged**